src/blame.ts

```diff
@@ -0,0 +1,82 @@
+import gql from 'tagged-template-noop'
+import sourcegraph from 'sourcegraph'
+
+/**
+ * Resolve a URI of the forms git://github.com/owner/repo?rev#path and file:///path to an absolute reference, using
+ * the given base (root) URI.
+ */
+export function resolveURI(uri: string): { repo: string; rev: string; path: string } {
+    const url = new URL(uri)
+    if (url.protocol === 'git:') {
+        return {
+            repo: (url.host + url.pathname).replace(/^\/*/, '').toLowerCase(),
+            rev: url.search.slice(1).toLowerCase(),
+            path: url.hash.slice(1),
+        }
+    }
+    throw new Error(`unrecognized URI: ${JSON.stringify(uri)} (supported URI schemes: git)`)
+}
+
+export interface Hunk {
+    startLine: number
+    endLine: number
+    author: {
+        person: {
+            email: string
+            displayName: string
+            user: {
+                username: string
+            } | null
+        }
+        date: string
+    }
+    rev: string
+    message: string
+    commit: {
+        url: string
+    }
+}
+
+export const queryBlameHunks = async (uri: string): Promise<Hunk[]> => {
+    const { repo, rev, path } = resolveURI(uri)
+    const { data, errors } = await sourcegraph.commands.executeCommand(
+        'queryGraphQL',
+        gql`
+            query GitBlame($repo: String!, $rev: String!, $path: String!) {
+                repository(name: $repo) {
+                    commit(rev: $rev) {
+                        blob(path: $path) {
+                            blame(startLine: 0, endLine: 0) {
```

Comment:
**thought**: So does this only take whoeever changed the first line in the file? Shouldn't it better aggregate authors in promixity of the cursor, or from the whole file?
If it's not related to the current line but to the file, it could also be as a panel view (but probably not worth investing time in this now).

Reply:
I think this [query gets the author of every line in the file](https://sourcegraph.com/api/console#%7B%22query%22%3A%22%23%20Type%20queries%20here%2C%20with%20completion%2C%20validation%2C%20and%20hovers.%5Cn%23%5Cn%23%20Here's%20an%20example%20query%20to%20get%20you%20started%3A%5Cn%5Cnquery%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20repository(name%3A%20%5C%22github.com%2Fsourcegraph%2Fsourcegraph%5C%22)%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20commit(rev%3A%20%5C%22main%5C%22)%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20blob(path%3A%20%5C%22client%2Fweb%2Fsrc%2FSourcegraphWebApp.tsx%5C%22)%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20blame(startLine%3A%200%2C%20endLine%3A%200)%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20startLine%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20endLine%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20author%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20person%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20email%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20displayName%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20user%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20username%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20date%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20message%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20rev%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20commit%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20url%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%7D%5Cn%5Cn%22%7D), and @marekweb creates a list of unique authors in `getAuthorsOfSurroundingCode`. It would be cool to weigh "knowledge" by proximity to the selected line(s).

Reply:
Yeah, looking up references would be the "intelligent" way to go about it. However, how would we deal with lines with nothing (or multiple things) to find references for, and for false positive references when using basic code intelligence?

What would a quick (~0.5d) solution look like?